javascript/ql/src/Performance/ReDoS.ql

```diff
@@ -639,8 +639,7 @@ newtype TStatePair =
   MkStatePair(State q1, State q2) {
     isFork(q1, _, _, _, _) and q2 = q1
     or
-    step(_, _, _, q1, q2) and
-    q1.toString() <= q2.toString()
+    step(_, _, _, q1, q2)
```

Comment:
Nice catch, that can't have been fun to debug. 😬 

Merging symmetric states in a product automaton is a legit optimization, though, so removing that doesn't seem right.

I think the real bug was that we forgot to check for a transition to (q2,q1) which is equal to (q1,q2) after merging symmetric states. Could you try this instead?
```codeql
(step(_, _, _, q1, q2) or step(_, _, _, q2, q1)) and
q1.toString() <= q2.toString()
```

Reply:
> Nice catch, that can't have been fun to debug. grimacing

Oh yeah! Debugging `predicate delta2(StatePair q, StatePair r)` was no fun. 

> Could you try this instead? 

That worked! And it slightly improves performance. 

I'll try to it generalizes it to 3-tuples (for super-linear ReDoS).